tests/UnifiedSpecTests/UnifiedSpecTest.php

```diff
@@ -0,0 +1,346 @@
+<?php
+
+namespace MongoDB\Tests\UnifiedSpecTests;
+
+use MongoDB\Client;
+use MongoDB\Driver\Exception\ServerException;
+use MongoDB\Driver\ReadPreference;
+use MongoDB\Driver\Server;
+use MongoDB\Operation\DatabaseCommand;
+use MongoDB\Tests\FunctionalTestCase;
+use stdClass;
+use Symfony\Bridge\PhpUnit\SetUpTearDownTrait;
+use Throwable;
+use function assertTrue;
+use function file_get_contents;
+use function glob;
+use function MongoDB\BSON\fromJSON;
+use function MongoDB\BSON\toPHP;
+use function sprintf;
+use function version_compare;
+
+/**
+ * Unified spec test runner.
+ *
+ * @see https://github.com/mongodb/specifications/pull/846
+ */
+class UnifiedSpecTest extends FunctionalTestCase
+{
+    use SetUpTearDownTrait;
+
+    const SERVER_ERROR_INTERRUPTED = 11601;
+
+    const TOPOLOGY_SINGLE = 'single';
+    const TOPOLOGY_REPLICASET = 'replicaset';
+    const TOPOLOGY_SHARDED = 'sharded';
+    const TOPOLOGY_SHARDED_REPLICASET = 'sharded-replicaset';
+
+    /** @var MongoDB\Client */
+    private static $internalClient;
+
+    /** @var FailPointObserver */
+    private $failPointObserver;
+
+    private static function doSetUpBeforeClass()
+    {
+        parent::setUpBeforeClass();
+
+        /* Provide internal client unmodified URI, since it may need to execute
+         * commands on multiple mongoses (e.g. killAllSessions) */
+        self::$internalClient = new Client(static::getUri(true));
+        self::killAllSessions();
+    }
+
+    private function doSetUp()
+    {
+        parent::setUp();
+
+        /* TODO: The transactions spec advises calling killAllSessions only at
+         * the start of the test suite and after failed tests; however, the
+         * "unpin after transient error within a transaction" pinning test
+         * causes the subsequent transaction test to block. This can be removed
+         * once that is addressed. */
+        self::killAllSessions();
+
+        $this->failPointObserver = new FailPointObserver();
+        $this->failPointObserver->start();
+    }
+
+    private function doTearDown()
+    {
+        if ($this->hasFailed()) {
+            self::killAllSessions();
+        }
+
+        $this->failPointObserver->stop();
+        $this->failPointObserver->disableFailPoints();
+
+        /* TODO: Consider manually invoking gc_collect_cycles since each test is
+         * prone to create cycles (perhaps due to EntityMap), which can leak and
+         * prevent sessions from being released back into the pool. */
```

Comment:
I only bumped into this while investigating the blocking transaction test issue, and that ultimately required `killAllSessions` as a workaround. I don't actually think this will have any effect on test execution, although there's some peace of mind in being able to trust that the EntityMap is totally dismantled. I imagine this will be more relevant when/if we start using Clients with persistence disabled, since we'd probably want to reap those after each test.

Reply:
Yep. Considering there already was some funkiness going on and with the segmentation fault appearing every once in a while due to the clusterTime being double-freed, I'm all good for using `gc_collect_cycles` as a precaution